python/ray/experimental/channel/shared_memory_channel.py

```diff
@@ -451,3 +458,94 @@ def close(self) -> None:
         if self.is_local_node(self._reader_node_id):
             self.ensure_registered_as_reader()
         self._worker.core_worker.experimental_channel_set_error(self._reader_ref)
+
+
+@PublicAPI(stability="alpha")
+class MultiChannel(ChannelInterface):
+    """
+    Can be used to send data to different readers via different channels.
+    For example, if the reader is in the same worker process as the writer,
+    the data can be sent via IntraProcessChannel. If the reader is in a different
+    worker process, the data can be sent via shared memory channel.
+
+    Args:
+        writer: The actor that may write to the channel. None signifies the driver.
+        readers: The actors that may read from the channel. None signifies
+            the driver.
+    """
+
+    def __init__(
+        self,
+        writer: Optional[ray.actor.ActorHandle],
+        readers: List[Optional[ray.actor.ActorHandle]],
+        _channel_dict: Optional[Dict[ray.ActorID, ChannelInterface]] = None,
+        _channels: Optional[List[ChannelInterface]] = None,
+    ):
+        self._writer = writer
+        self._readers = readers
+        # A dictionary that maps the actor ID to the channel object.
+        self._channel_dict = _channel_dict or {}
+        # A list of channel objects.
+        self._channels = _channels or []
```

Comment:
Why do we need both `self._channels` and `self._channel_dict`? If you need the channels as a list, you can do `self._channel_dict.values()`.

Reply:
We may have multiple readers read the same channel. For example,

```python
channel_dict = {actor1: channel1, actor2: channel2, actor3: channel1}

# We will write to `channel1` twice.
def write(self, value: Any):
    for channel in channel_dict.values():  # [channel1, channel2, channel1]
        channel.write(value)
```

Reply:
We will always know all of the readers up front, so it seems better to just use a dictionary to simplify the implementation.

Reply:
> We will always know all of the readers up front, so it seems better to just use a dictionary to simplify the implementation.

@jackhumphries would you mind adding more details about this? Do you mean that:

* Maintain the dict `self._channel_dict` (key: reader, value: channel).
* When we need to write, we do something like:
  ```python
  def write(self, value: Any):
    seen = set()
    for channel in self._channel_dict:
      if channel not in seen:
        channel.write(value)
        seen.add(channel)
  ```

Reply:
I think I may understand your points: https://github.com/ray-project/ray/pull/45591#discussion_r1631437749.

Reply:
Added